Add GeoLocationInput component

samshara opened the pull request:
- fixes the-deep/client#2106
- Depends on https://github.com/the-deep/server/pull/1008

## Changes

* Create GeoLocationInput component
* Use GeoLocationInput component to tag entry
* Add map based location selection

## This PR doesn't introduce any:

- [x] temporary files, auto-generated files or secret keys
- [x] build works
- [x] eslint issues
- [x] typescript issues
- [x] `console.log` meant for debugging
- [x] typos
- [x] unwanted comments
- [x] conflict markers

## This PR contains valid:

- [ ] permission checks
- [ ] translations

samshara committed:
Create GeoLocationInput component

Use GeoLocationInput component to tag location

AdityaKhatri commented on app/components/GeoLocationInput/GeoLocationModal/GeoLocationMapInput/GeoAreaListItem/index.tsx:
```diff
@@ -0,0 +1,48 @@
+        onDismiss,
+    } = props;
+
+    const handleDismissButtonClick = useCallback(
+        () => {
+            onDismiss(id);
+        },
+        [onDismiss, id],
+    );
+    return (
+        <div className={_cs(className, styles.geoAreaListItem)}>
+            <ElementFragments
+                actions={(
+                    <QuickActionButton
+                        name="delete"
+                        onClick={handleDismissButtonClick}
```
```suggestion
    return (
        <div className={_cs(className, styles.geoAreaListItem)}>
            <ElementFragments
                actions={(
                    <QuickActionButton
                        name={id}
                        onClick={onDismiss}
```

AdityaKhatri commented on app/components/GeoLocationInput/GeoLocationModal/index.tsx:
```diff
@@ -0,0 +1,85 @@
+        setTempGeoAreas,
+    ] = useState<string[] | undefined>();
+
+    useEffect(() => {
+        setTempGeoAreas(selectedGeoAreas ?? undefined);
+    }, [selectedGeoAreas]);
```
Can we not make this the default value instead of using an effect?

tnagorra replied:
We should use default value.

But, we may still need this logic to override tesmpGeoAreas from parent.
@samshara ?

samshara replied:
lifted the state to parent

AdityaKhatri commented:
While switching between multiple regions within the project, the default selection of admin-level should be set to admin 0 if possible.

AdityaKhatri committed:
Fix review comments in geo location widget

AdityaKhatri commented:
@samshara , if you select some regions and save it and refresh the page, you do not see the name of selected options.
![image](https://user-images.githubusercontent.com/7893887/140706354-607dc520-de1c-46b8-9b7a-7acb7b6b7c8d.png)

tnagorra commented on app/components/GeoMultiSelectInput/index.tsx:
```diff
@@ -30,7 +30,7 @@ const GEOAREAS = gql`
 
 const keySelector = (d: GeoArea) => d.id;
-const labelSelector = (d: GeoArea) => d.title;
+const labelSelector = (d: GeoArea) => `${d.regionTitle}/${d.adminLevelTitle}/${d.title}`;
```
I propose we use '›' instead of '/' as separator?

tnagorra commented on app/components/GeoLocationInput/GeoLocationModal/GeoLocationMapInput/GeoLocationMap/index.tsx:
```diff
@@ -0,0 +1,299 @@
+const scaleFactor = 1;
+const sourceOptions: AnySourceData = {
+    type: 'geojson',
+    // FIXME: we need to think about this
```
Let's remove this FIXME comment

tnagorra commented on app/components/GeoLocationInput/GeoLocationModal/GeoLocationMapInput/GeoLocationMap/index.tsx:
```diff
@@ -0,0 +1,299 @@
+        ],
+        'fill-color': ['case',
+            ['boolean', ['feature-state', 'selected'], false],
+            '#1a3ed0',
+            '#a9bedc',
```
Do we need to create constants for maps in deep? @frozenhelium @samshara @AdityaKhatri

samshara replied:
@AdityaKhatri any thoughts?

tnagorra commented on app/components/GeoLocationInput/GeoLocationModal/GeoLocationMapInput/GeoLocationMap/index.tsx:
```diff
@@ -0,0 +1,299 @@
+import React, { useMemo, useState, useCallback } from 'react';
```
We should have used the existing geolocation map component

tnagorra commented on app/components/GeoLocationInput/GeoLocationModal/GeoLocationMapInput/index.tsx:
```diff
@@ -0,0 +1,250 @@
+            variables,
+            onCompleted: (data) => {
+                const [topRegion] = data.project?.regions ?? [];
+                const [topAdminLevel] = topRegion.adminLevels ?? [];
```
What if `topRegion` is undefined?

samshara replied:
if we don't have topRegion no region will be selected.

tnagorra commented on app/components/GeoLocationInput/GeoLocationModal/GeoLocationMapInput/index.tsx:
```diff
@@ -0,0 +1,250 @@
+        },
+    );
+
+    const adminLevels = projectRegions?.project?.regions
+        ?.filter((v) => v.id === selectedRegion)
+        .map((r) => r.adminLevels)
+        .flat()
+        .filter(isDefined);
```
Any reason we are not memoizing this?

tnagorra commented on app/components/GeoLocationInput/GeoLocationModal/GeoLocationMapInput/index.tsx:
```diff
@@ -0,0 +1,250 @@
+
+    const geoAreasRendererParams = useCallback((_: string, geoArea: GeoArea) => ({
+        id: `${geoArea.id}`,
+        value: `${geoArea.regionTitle}/${geoArea.adminLevelTitle}/${geoArea.title}`,
```
Let's add a simple function to concatenate geolocation names as breadcrumbs.

tnagorra commented on app/components/GeoLocationInput/GeoLocationModal/GeoLocationMapInput/index.tsx:
```diff
@@ -0,0 +1,250 @@
+    }), [handleRemoveItem]);
+
+    const geoAreasList = useMemo(() => (
+        tempGeoAreas
+            ?.map((val: string) => geoAreaOptions?.find((v) => v.id === val))
+            .filter(isDefined)
```
Let's compute this the other way round.
```javascript
geoAreasOptions.filter((item) => tempGeoAreas.includes(item))
```

The code complexity here is O(N^2), if the tempGeoAreas can be large
We should considering using a Map instead

AdityaKhatri committed:
Add compact geo location input

AdityaKhatri committed:
Use common region component for view and tagging purpose

samshara committed:
Minor fixes in GeoLocationInput component

* Show values after refresh
* Select top admin level ( level = 0) when new region selection

tnagorra commented on app/views/Project/Tagging/Sources/components/EditableEntry/index.tsx:
```diff
@@ -86,6 +87,8 @@ interface Props {
     compact?: boolean;
     entryImage: Entry['image'] | undefined | null;
     onEntryDataChange: () => void;
+    geoAreas: GeoArea[] | undefined | null;
+    onGeoAreasChange: React.Dispatch<React.SetStateAction<GeoArea[] | undefined | null>>;
```
```suggestion
    geoAreaOptions: GeoArea[] | undefined | null;
    onGeoAreaOptionsChange: React.Dispatch<React.SetStateAction<GeoArea[] | undefined | null>>;
```

tnagorra commented on app/views/Project/Tagging/Sources/SourcesGrid/EntryCard/index.tsx:
```diff
@@ -64,6 +65,8 @@ interface Props {
     onViewTagsButtonClick?: (entryId: string) => void;
     onHideTagsButtonClick?: (entryId: string) => void;
     onEntryDataChange: () => void;
+    geoAreas: GeoArea[] | undefined | null;
+    onGeoAreasChange: React.Dispatch<React.SetStateAction<GeoArea[] | undefined | null>>;
```
Let's rename geoAreas to geoAreaOptions

tnagorra commented on app/components/region/RegionMap/index.tsx:
```diff
@@ -86,12 +114,14 @@ interface GeoAreaBoundsResponse {
     title?: string;
     navigationDisabled?: boolean;
+    selectedGeoAreas?: string[] | null ;
+    onGeoAreasSelectionChange?: (value: string[]) => void;
```
```suggestion
    onGeoAreasChange?: (value: string[]) => void;
```

tnagorra commented on app/components/framework/CompactAttributeInput/GeoLocationWidgetInput/index.tsx:
```diff
@@ -0,0 +1,83 @@
+    readOnly?: boolean;
+
+    widget: PartialGeoLocationWidget;
+    geoAreas: GeoArea[] | undefined | null;
+    onGeoAreasChange: React.Dispatch<React.SetStateAction<GeoArea[] | undefined | null>>;
```
```suggestion
    geoAreaOptions: GeoArea[] | undefined | null;
    onGeoAreaOptionsChange: React.Dispatch<React.SetStateAction<GeoArea[] | undefined | null>>;
```

tnagorra commented on app/components/framework/AttributeInput/GeoLocationWidgetInput/index.tsx:
```diff
@@ -0,0 +1,86 @@
+
+    const onChange = useCallback(
+        (val: GeoLocationValue['value'] | undefined, inputName: N) => {
+            if (isNotDefined(val)) {
+                onChangeFromProps(undefined, inputName);
+            } else {
+                onChangeFromProps({ value: val }, inputName);
+            }
```
Let's add a TODO to handle points and polygons

tnagorra commented on app/components/entry/Section/index.tsx:
```diff
@@ -20,6 +21,8 @@ interface Props {
     attributesMap: Partial<Record<string, { index: number, value: WidgetAttribute }>>;
     readOnly?: boolean;
     disabled?: boolean;
+    geoAreas: GeoArea[] | undefined | null;
+    onGeoAreasChange: React.Dispatch<React.SetStateAction<GeoArea[] | undefined | null>>;
```
```suggestion
    geoAreaOptions: GeoArea[] | undefined | null;
    onGeoAreaOptionsChange: React.Dispatch<React.SetStateAction<GeoArea[] | undefined | null>>;
```

tnagorra commented on app/components/entry/EntryInput/index.tsx:
```diff
@@ -60,6 +61,8 @@ interface EntryInputProps<T extends string | number | undefined> {
     compact?: boolean;
 
     entryImage: Entry['image'] | undefined | null;
+    geoAreas: GeoArea[] | undefined | null;
+    onGeoAreasChange: React.Dispatch<React.SetStateAction<GeoArea[] | undefined | null>>;
```
```suggestion
    geoAreaOptions: GeoArea[] | undefined | null;
    onGeoAreaOptionsChange: React.Dispatch<React.SetStateAction<GeoArea[] | undefined | null>>;
```

tnagorra commented on app/components/entry/CompactSection/index.tsx:
```diff
@@ -33,6 +34,8 @@ export interface Props {
     entryClientId: string;
     sectionId?: string;
     onAddButtonClick: (entryId: string, sectionId?: string) => void;
+    geoAreas: GeoArea[] | undefined | null;
+    onGeoAreasChange: React.Dispatch<React.SetStateAction<GeoArea[] | undefined | null>>;
```
```suggestion
    geoAreaOptions: GeoArea[] | undefined | null;
    onGeoAreaOptionsChange: React.Dispatch<React.SetStateAction<GeoArea[] | undefined | null>>;
```

tnagorra commented on app/components/GeoMultiSelectInput/index.tsx:
```diff
@@ -30,7 +30,7 @@ const GEOAREAS = gql`
 
 const keySelector = (d: GeoArea) => d.id;
-const labelSelector = (d: GeoArea) => d.title;
+const labelSelector = (d: GeoArea) => `${d.regionTitle}>${d.adminLevelTitle}>${d.title}`;
```
Let's use chevron instead of '>'

tnagorra commented on app/components/GeoLocationInput/GeoLocationModal/index.tsx:
```diff
@@ -0,0 +1,70 @@
+    className?: string;
+    projectId: string;
+    onSubmit: () => void;
+    tempGeoAreas?: string[] | null | undefined;
+    geoAreaOptions: GeoArea[] | null | undefined;
+    onTempGeoAreasChange: React.Dispatch<React.SetStateAction<string[] | null | undefined>>;
```
What is tempGeoAreas?

tnagorra commented on app/components/GeoLocationInput/GeoLocationModal/GeoLocationMapInput/index.tsx:
```diff
@@ -0,0 +1,239 @@
+interface Props {
+    className?: string;
+    projectId: string;
+    tempGeoAreas?: string[] | null | undefined;
```
What is tempGeoAreas?

AdityaKhatri committed:
Rename variable and callbacks in region maps

tnagorra committed:
Clear form data on cancel (editable entry)

- Improve styling for Geolocation Filter Input
- Improve styling for Geolocation Attribute Input

tnagorra committed:
Set fixed width for filters
